### PR TITLE
VMware: vmware_guest_facts: add the ability to get guest facts and specific properties using the vSphere schema

### PR DESCRIPTION
**Pull request by jeking3**
##### SUMMARY

The module `vmware_guest_facts` mutates the vSphere results into a form that appears to align with the result from the libvirt guest facts.  This output is a subset of the total information available about the guest, and the output format is [not documented (see results -> instance)](https://docs.ansible.com/ansible/2.7/modules/vmware_guest_facts_module.html).

This pull request adds the ability to optionally specify the output format of the fact gathering to be in one of the following two formats:
- `summary` is the existing format in the module, and the default
- `vsphere` is the vSphere API documented format, in JSON form

The properties are an optional list of dotted property names, identical to the format the property collector would normally take.  When a property list is given, those properties are extracted from the vim object that results from the standard `get_vm` lookup that was already there.  (The property list is not passed into the property collector - that would have required modifying the vm lookup code paths which are already complex).

An example configuration file for the module is:
```
{
  "ANSIBLE_MODULE_ARGS": {
    "hostname": "0.0.0.0",
    "username": "myuser",
    "password": "mypass",
    "validate_certs": false,
    "datacenter": "MyDatacenter",
    "name": "MyVM",
    "format": "vsphere",
    "properties": [
      "config.hardware.memoryMB",
      "config.hardware.numCPU",
      "overallStatus",
      "name",
      "guest.disk"
    ]
  }
}
```
Example result:
```
{
  "instance": {
    "config": {
      "hardware": {
        "memoryMB": 4096,
        "numCPU": 2
      }
    },
    "overallStatus": "green",
    "name": "MyVM",
    "guest": {
      "disk": [
        {
          "freeSpace": 7178031104,
          "diskPath": "C:\\",
          "capacity": 20880289792
        }
      ]
    }
  }
}
```

Advantages of vSphere API access are:

1. The structure of the results is [well-documented](https://code.vmware.com/apis/358/vsphere).
2. All of the properties are available.
3. One can select any proper subset of properties (including all of them).

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME

vmware_guest_facts

##### ANSIBLE VERSION

```paste below
ansible 2.8.0.dev0 (enhanced-vm-facts d2f18f3011) last updated 2018/10/22 19:12:45 (GMT +000)
  config file = None
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /ansible/lib/ansible
  executable location = /ansible/bin/ansible
  python version = 2.7.15 (default, Sep 12 2018, 02:38:23) [GCC 6.4.0]
```

##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

**Comment by jeking3** [comment]
This is also blocked on an upstream pyVmomi PR: https://github.com/vmware/pyvmomi/pull/732

**Comment by Akasurde** [comment]
@jeking3 Thanks for the PR, do you think we should add a version check for Pyvmomi as this version will land up in 6.7.* or something ? Thanks

**Comment by jeking3** [comment]
@Akasurde I'd say we need to wait for the upstream merge to determine what that will end up being.  It would probably be a good idea to roll this out across all the fact gathering modules, or what I will likely eventually do is make a vmware_facts module that can extract any or all properties of any type supported by the property collector.  That way one only needs a single facts module for all of vmware and it returns a well-documented schema in the vSphere API.

**Comment by Akasurde** [comment]
@jeking3 make sense to me. Can you also take a look at [this](https://github.com/ansible/ansible/blob/82f1438b14a9753d42d03c2bbda96c5adc301dc5/lib/ansible/module_utils/vmware.py#L614) ?

**Comment by jeking3** [comment]
Something like `serialize_spec` would be likely be unnecessary with the `VmomiJSONEncoder`.  So using `managed_object_to_json` which is in the pull request would be my choice... of course once pyvmomi merges and releases that is...

**Comment by jeking3** [comment]
Perhaps it would be best to leave all the individual `vmware_*_facts` as they are and just add `vmware_facts` using the same folder-based resolver that's already there, and removing the ambiguity that exists when retrieving things.  (i.e. "name_match: First" is evil, should be an error on ambiguity, and remove calls the FindBy*(singular) and replace with FindAllBy* and test for at most one result.

**Comment by jeking3** [comment]
Waiting for pyvmomi to issue a new release with the code necessary; then can change the dependencies here to match.  It will be something like `pyvmomi>6.7.1`

**Comment by jeking3** [comment]
Okay, since I don't know when pyVmomi is going to create another release (they JUST made one, but it doesn't include the upstream support) I modified the module to fail if folks do not have the `VmomiSupport.VmomiJSONEncoder` in their pyVmomi.  As such, if this passes tests then it is ready to be included.

File: lib/ansible/module_utils/vmware.py
Comment (Akasurde) [inline lib/ansible/module_utils/vmware.py:1293]:
```
result = dict()
```

File: lib/ansible/module_utils/vmware.py
Comment (Akasurde) [inline lib/ansible/module_utils/vmware.py]:
```suggestion
                except (AttributeError, KeyError):
```

**Review by Akasurde — CHANGES_REQUESTED** [review]
Rest LGTM, Could you please add test case for this change ?

**Comment by Akasurde** [comment]
@jeking3 Any news ?

**Comment by jeking3** [comment]
Hi, sorry, there have been other projects consuming my time.  I will get back to this, it is on my list.

**Comment by jeking3** [comment]
It won't be possible to test this code unless we pull in an unreleased git commit from the pyvmomi project, so I left it testing for the failure case when you don't have the commit you need to convert to json.  Once upstream releases something > 6.7.1 then I will swing back around and enable this.

**Comment by jeking3** [comment]
I've asked the pyvmomi team if they can release a 6.7.1.2018.12 release that includes the bits we need.

File: lib/ansible/module_utils/vmware.py
Comment (Akasurde) [reply lib/ansible/module_utils/vmware.py:1293]: @jeking3 Could you please address this ?

**Comment by Akasurde** [comment]
@jeking3 Ping me once you are done. I will take it forward.

**Comment by jeking3** [comment]
The pyvmomi team has released version v6.7.1.2018.12 which includes the necessary upstream code to perform the JSON encoding that this new code depends on, so I am going to complete this PR now.  I'll let you know when that is done.

**Comment by Akasurde** [comment]
@jeking3 Thanks let me know once your done. I will do final review.

**Comment by Akasurde** [comment]
@pdellaert @dericcrago @Im0 @tchernomax @ckotte Could you please review this ?

**Comment by jeking3** [comment]
Need some help from folks who know the integration test.  I added a requirements.txt to the integration test for vmware_guest_facts and it was ignored as far as I can tell.  I said: {{ pyvmomi>6.7.1 }} but the code failed indicating pyVmomi.VmomiJSONEncoder was not present.

**Comment by jeking3** [comment]
Oh, nevermind I see what the fortios folks did now. :)

**Comment by Akasurde** [comment]
@jeking3 Are you editing this file -  https://github.com/ansible/ansible/blob/devel/test/runner/requirements/integration.cloud.vcenter.txt ?

**Comment by jeking3** [comment]
@Akasurde I did not want to force pyvmomi>6.7.1 on all vmware testing so I am doing that the fortios folks did in their integration test.  I just panicked too quickly and asked for help... sorry!

**Comment by jeking3** [comment]
Empty force push, docker environmental build issue could not reach quay.

File: test/integration/targets/vmware_guest_facts/tasks/main.yml
Comment (mattclay) [inline test/integration/targets/vmware_guest_facts/tasks/main.yml]: Instead of installing the requirements here, add a version constraint for all the tests to use:

https://github.com/ansible/ansible/blob/devel/test/runner/requirements/constraints.txt

**Review by mattclay — DISMISSED** [review]

**Comment by jeking3** [comment]
Sadly, vcsim appears to be incompatible with this, even though vCenter is not.
```
TASK [vmware_guest_facts : Get all details about virtual machines using the vsphere output schema] ***
04:03 task path: /root/ansible/test/integration/targets/vmware_guest_facts/tasks/main.yml:106
04:03 <127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: root
04:03 <127.0.0.1> EXEC /bin/sh -c 'echo ~root && sleep 0'
04:03 <127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /root/.ansible/tmp/ansible-tmp-1545323163.24-139565755192520 `" && echo ansible-tmp-1545323163.24-139565755192520="` echo /root/.ansible/tmp/ansible-tmp-1545323163.24-139565755192520 `" ) && sleep 0'
04:03 Using module file /root/ansible/lib/ansible/modules/cloud/vmware/vmware_guest_facts.py
04:03 <127.0.0.1> PUT /root/.ansible/tmp/ansible-local-10440nrDNp6/tmp3dOUPD TO /root/.ansible/tmp/ansible-tmp-1545323163.24-139565755192520/AnsiballZ_vmware_guest_facts.py
04:03 <127.0.0.1> EXEC /bin/sh -c 'chmod u+x /root/.ansible/tmp/ansible-tmp-1545323163.24-139565755192520/ /root/.ansible/tmp/ansible-tmp-1545323163.24-139565755192520/AnsiballZ_vmware_guest_facts.py && sleep 0'
04:03 <127.0.0.1> EXEC /bin/sh -c '/tmp/ansible-test-coverage-o2s_a62i/coverage/injector.py /root/.ansible/tmp/ansible-tmp-1545323163.24-139565755192520/AnsiballZ_vmware_guest_facts.py && sleep 0'
04:03 <127.0.0.1> EXEC /bin/sh -c 'rm -f -r /root/.ansible/tmp/ansible-tmp-1545323163.24-139565755192520/ > /dev/null 2>&1 && sleep 0'
04:03 The full traceback is:
04:03 WARNING: The below traceback may *not* be related to the actual failure.
04:03   File "/tmp/ansible_vmware_guest_facts_payload_dOdz3r/__main__.py", line 244, in main
04:03     instance = pyv.to_json(vm, module.params['properties'])
04:03   File "/tmp/ansible_vmware_guest_facts_payload_dOdz3r/ansible_vmware_guest_facts_payload.zip/ansible/module_utils/vmware.py", line 1267, in to_json
04:03     result = self._jsonify(obj)
04:03   File "/tmp/ansible_vmware_guest_facts_payload_dOdz3r/ansible_vmware_guest_facts_payload.zip/ansible/module_utils/vmware.py", line 1230, in _jsonify
04:03     sort_keys=True, strip_dynamic=True))
04:03   File "/usr/lib/python2.7/json/__init__.py", line 251, in dumps
04:03     sort_keys=sort_keys, **kw).encode(obj)
04:03   File "/usr/lib/python2.7/json/encoder.py", line 209, in encode
04:03     chunks = list(chunks)
04:03   File "/usr/lib/python2.7/json/encoder.py", line 442, in _iterencode
04:03     o = _default(o)
04:03   File "/usr/local/lib/python2.7/dist-packages/pyVmomi/VmomiSupport.py", line 357, in default
04:03     result[prop.name] = getattr(obj, prop.name)
04:03   File "/usr/local/lib/python2.7/dist-packages/pyVmomi/VmomiSupport.py", line 700, in __call__
04:03     return self.f(*args, **kwargs)
04:03   File "/usr/local/lib/python2.7/dist-packages/pyVmomi/VmomiSupport.py", line 520, in _InvokeAccessor
04:03     return self._stub.InvokeAccessor(self, info)
04:03   File "/usr/local/lib/python2.7/dist-packages/pyVmomi/StubAdapterAccessorImpl.py", line 42, in InvokeAccessor
04:03     options=self._pcType.RetrieveOptions(maxObjects=1))
04:03   File "/usr/local/lib/python2.7/dist-packages/pyVmomi/VmomiSupport.py", line 706, in <lambda>
04:03     self.f(*(self.args + (obj,) + args), **kwargs)
04:03   File "/usr/local/lib/python2.7/dist-packages/pyVmomi/VmomiSupport.py", line 512, in _InvokeMethod
04:03     return self._stub.InvokeMethod(self, info, args)
04:03   File "/usr/local/lib/python2.7/dist-packages/pyVmomi/SoapAdapter.py", line 1365, in InvokeMethod
04:03     raise exc
04:03 
04:03 fatal: [testhost]: FAILED! => {
04:03     "changed": false, 
04:03     "invocation": {
04:03         "module_args": {
04:03             "datacenter": "DC0", 
04:03             "folder": null, 
04:03             "hostname": "vcenter-simulator", 
04:03             "name": null, 
04:03             "name_match": "first", 
04:03             "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", 
04:03             "port": 443, 
04:03             "properties": null, 
04:03             "schema": "vsphere", 
04:03             "tags": false, 
04:03             "username": "user", 
04:03             "uuid": "9899b30c-bb3d-4631-b30d-2e021957bb86", 
04:03             "validate_certs": false
04:03         }
04:03     }, 
04:03     "msg": "Fact gather failed with exception 'xml document: <httplib.HTTPResponse instance at 0x7f40c0d9dea8> parse error at: line:2, col:524'"
04:03 }
04:03 	to retry, use: --limit @/root/ansible/test/integration/vmware_guest_facts-4zamztxf.retry
```

I'll try to figure it out.  What a drag...

File: test/integration/targets/vmware_guest_facts/tasks/main.yml
Comment (jeking3) [reply test/integration/targets/vmware_guest_facts/tasks/main.yml]: Thanks, I changed it.

**Comment by jeking3** [comment]
Did a force push to fix environmental build issue with quay.

**Comment by jeking3** [comment]
Well some good news, I can reproduce the issue with the simulator locally.

**Comment by jeking3** [comment]
Does ansible have to support Python 2.6?  From [PEP 361](https://www.python.org/dev/peps/pep-0361/):
```
With its release on October 29, 2013, all official support for Python 2.6 has ended. Python 2.6 is no longer being maintained for any purpose.
```
Until ansible drops python 2.6 support, this cannot be merged.  It requires pyVmomi > 6.7.1.

**Comment by mattclay** [comment]
@jeking3 Features that have dependencies which require newer Python versions can sometimes still be supported. Those features need to be optional, available only when the newer dependency is present, with graceful handling when it is not. The documentation will need to reflect the additional requirements for the optional features.

**Comment by jeking3** [comment]
Okay, so then I need to change it to detect whether pyVmomi has the support it needs, and if not then error out with a message indicating pyVmomi must be > 6.7.1.

**Comment by mattclay** [comment]
@jeking3 Yes, as long as the error only occurs when trying to use the new features that require pyVmomi > 6.7.1. If using existing features, older versions of pyVmomi should continue to work.

You will also need to add to the existing test constraints:

https://github.com/ansible/ansible/blob/6ac22416af987b4cd3e4cadf0f1d01352daf203c/test/runner/requirements/constraints.txt#L31

On the line below that add:

```
pyvmomi > 6.7.1 ; python_version >= '2.7' # for pyVmomi conversion to JSON support
```

**Comment by jeking3** [comment]
What you suggested ensures pyVmomi > 6.7.1 on all systems with python 2.7 or later.  I'm okay with that if you are.

**Comment by mattclay** [comment]
@jeking3 Will there be any integration tests covering the new features? If so, then we'll need the upgraded version. If not, then don't update the constraints.

**Comment by jeking3** [comment]
Sorry, I get it now - that's just for the integration test.  I had it in my mind that was for ansible runtime.  Thanks.

**Comment by mattclay** [comment]
One other thought, if we're updating the constraint, is that we only run the integration tests on Python 2.7 and 3.6 currently. To make sure the code still works under Python 2.6 with an older version of pyvmomi we probably need at least a small unit test to check that. Alternatively we could enable the integration tests for Python 2.6.

**Comment by jeking3** [comment]
Perhaps I wasn't clear before, but continuing to support Python 2.6 is a waste of energy.
https://www.python.org/dev/peps/pep-0361/
Laat release October 2013, support (including security patches) no longer happening.

**Comment by jeking3** [comment]
This is now passing along with an integration test.

**Comment by mattclay** [comment]
@jeking3 While Ansible has dropped official support for Python 2.6 on the controller, management of systems with Python 2.6 is still officially supported. Although the Python Software Foundation no longer supports Python 2.6, there are currently supported Linux distributions which include Python 2.6. Testing is an important part of support, which is why we're concerned about testing modules on Python 2.6.

**Comment by jeking3** [comment]
@Akasurde this is ready for final review.

**Review by Akasurde — APPROVED** [review]

**Comment by Akasurde** [comment]
@mattclay Are you OK in merging this ? or do you have any comment which are unanswered ? Thanks.

**Review by pdellaert — APPROVED** [review]

**Comment by Akasurde** [comment]
rebuild_merge

**Comment by Akasurde** [comment]
@jeking3 Thanks for the contribution. @pdellaert @mattclay Thanks for the review.

**Pull request merged**
